verl/experimental/agent_loop/tool_agent_loop.py

```diff
@@ -91,139 +200,170 @@ async def run(self, sampling_params: dict[str, Any], **kwargs) -> AgentLoopOutpu
-                    None,
-                    lambda messages=tool_messages: self.tokenizer.apply_chat_template(
-                        messages, add_generation_prompt=True, tokenize=True, **self.apply_chat_template_kwargs
-                    ),
-                )
-            tool_response_ids = tool_response_ids[len(self.system_prompt) :]
 
-            # NOTE: last turn should not be user turn, or the EOS token reward
-            # can't be propagated to previous token in GAE.
-            if len(response_mask) + len(tool_response_ids) >= self.response_length:
-                break
+        # Update prompt with tool responses
+        if self.processor is not None:
+            raw_tool_response = await self.loop.run_in_executor(
+                None,
+                lambda messages=add_messages: self.processor.apply_chat_template(
+                    messages, add_generation_prompt=True, tokenize=False, **self.apply_chat_template_kwargs
+                ),
+            )
+            # Use only the new images from this turn for processing tool responses
+            current_images = new_images_this_turn if new_images_this_turn else None
+            model_inputs = self.processor(text=[raw_tool_response], images=current_images, return_tensors="pt")
+            response_ids = model_inputs.pop("input_ids").squeeze(0).tolist()
+        else:
+            response_ids = await self.loop.run_in_executor(
+                None,
+                lambda messages=add_messages: self.tokenizer.apply_chat_template(
+                    messages, add_generation_prompt=True, tokenize=True
+                ),
+            )
+        response_ids = response_ids[len(self.system_prompt) :]
 
-            prompt_ids += tool_response_ids
-            response_mask += [0] * len(tool_response_ids)
-            if response_logprobs:
-                response_logprobs += [0.0] * len(tool_response_ids)
-            user_turns += 1
+        # Update prompt_ids and response_mask
+        prompt_ids += response_ids
+        response_mask += [0] * len(response_ids)
+        if response_logprobs:
+            response_logprobs += [0.0] * len(response_ids)
 
-        response_ids = prompt_ids[-len(response_mask) :]
-        prompt_ids = prompt_ids[: len(prompt_ids) - len(response_mask)]
+        return AgentState.GENERATING, responses
 
-        multi_modal_data = {"image": image_data} if image_data is not None else {}
+    async def _handle_interacting_state(self, interaction, request_id, messages, interaction_kwargs, add_messages):
+        """Handle the interacting state: get user input from interaction."""
+        (
+            should_terminate_sequence,
+            interaction_responses,
+            reward,
+            metrics,
+        ) = await interaction.generate_response(request_id, messages, **interaction_kwargs)
+        responses = [{"role": "user", "content": interaction_responses}]
```

Comment:
Interaction response should be encoded to token ids and appended to prompt_ids, just like how we process the `[{"role": "tool", "content": "..."}` in `_handle_processing_tools_state `.